master/internal/provisioner/scale_decider.go

```diff
@@ -149,6 +149,10 @@ func (s *scaleDecider) findInstancesToTerminate(
 		toTerminate[id] = true
 	}
 
+	// TODO: Why do we terminate idle instances here? Does is indicate an error of some
```

Comment:
@shiyuann do you know the answer to this?

Reply:
it's just to reduce resource usage.

> should we prioritize shutting down spot requests first?

spot requests are very similar to starting instances but we should handle them differently from starting instances, which I saw you handled well in this PR using `SpotRequestPendingAWS` as a separate state.

we used not to prioritize shutting down starting instances because the past provisioner is based on an assumption that the current tasks cannot be scheduled on the current agents so shutting down starting instances might not be useful. But in the new architecture, we assume each resource pool only has one resource type (although this assumption might break for on-premise) so we might be able to prioritize shutting down starting instances after we introduce resource pools.